packages/kolme/src/core/execute.rs

```diff
@@ -198,22 +198,32 @@ impl<App: KolmeApp> ExecutionContext<'_, App> {
         chain: ExternalChain,
         event: &BridgeEvent,
         event_id: BridgeEventId,
-    ) -> Result<()> {
-        anyhow::ensure!(self
+    ) -> std::result::Result<(), KolmeError> {
+        if !self
             .framework_state
             .get_validator_set()
             .listeners
-            .contains(&self.pubkey));
+            .contains(&self.pubkey)
+        {
+            return Err(KolmeError::NotInValidatorSet {
+                signer: Box::new(self.pubkey),
+                role: "listener".to_string(),
+            });
+        }
 
         let state = self.framework_state.chains.get_mut(chain)?;
 
         let attestations = match state.pending_events.get_mut(&event_id) {
             Some(pending) => {
-                anyhow::ensure!(pending.event == *event);
+                if pending.event != *event {
+                    return Err(KolmeError::Other("Mismatched bridge event".to_string()));
```

Comment:
Any reason to fall back to `Other` here? I think it might make sense to introduce a module-level error type for execution, e.g. `KolmeExecutionError`, and then provide a variant on `KolmeError` like `enum KolmeError { ..., Execution(KolmeExecutionError) }`.